leshan-core/src/main/java/org/eclipse/leshan/core/node/TimestampedLwM2mNodesImpl.java

```diff
@@ -0,0 +1,131 @@
+/*******************************************************************************
+ * Copyright (c) 2021 Orange.
+ *
+ * All rights reserved. This program and the accompanying materials
+ * are made available under the terms of the Eclipse Public License v2.0
+ * and Eclipse Distribution License v1.0 which accompany this distribution.
+ *
+ * The Eclipse Public License is available at
+ *    http://www.eclipse.org/legal/epl-v20.html
+ * and the Eclipse Distribution License is available at
+ *    http://www.eclipse.org/org/documents/edl-v10.html.
+ *
+ * Contributors:
+ *     Orange - Send with multiple-timestamped values
+ *******************************************************************************/
+package org.eclipse.leshan.core.node;
+
+import java.util.HashMap;
+import java.util.Map;
+import java.util.Set;
+import java.util.TreeMap;
+
+/**
+ * The default implementation of {@link TimestampedLwM2mNodes}
+ */
+public class TimestampedLwM2mNodesImpl implements TimestampedLwM2mNodes {
+
+    private final Map<Long, Map<LwM2mPath, LwM2mNode>> timestampedPathNodesMap = new TreeMap<>((o1, o2) -> {
+        if (o1 == null) {
+            return (o2 == null) ? 0 : -1;
+        } else if (o2 == null) {
+            return 1;
+        } else {
+            return o1.compareTo(o2);
+        }
+    });
+
+    public TimestampedLwM2mNodesImpl() {
+    }
+
+    public TimestampedLwM2mNodesImpl(Map<LwM2mPath, LwM2mNode> pathNodesMap) {
+        timestampedPathNodesMap.put(null, pathNodesMap);
+    }
+
+    public TimestampedLwM2mNodesImpl(Long timestamp, LwM2mPath path, LwM2mNode node) {
+        if (!timestampedPathNodesMap.containsKey(timestamp)) {
+            timestampedPathNodesMap.put(timestamp, new HashMap<>());
+        }
+        timestampedPathNodesMap.get(timestamp).put(path, node);
+    }
+
+    @Override
+    public Map<Long, Map<LwM2mPath, LwM2mNode>> getTimestampedNodes() {
+        return timestampedPathNodesMap;
+    }
+
+    @Override
+    public Map<LwM2mPath, LwM2mNode> getNodesForTimestamp(Long timestamp) {
+        return timestampedPathNodesMap.get(timestamp);
+    }
+
+    @Override
+    public Map<LwM2mPath, LwM2mNode> getNodes() {
+        Map<LwM2mPath, LwM2mNode> result = new HashMap<>();
+        for (Map.Entry<Long, Map<LwM2mPath, LwM2mNode>> entry : timestampedPathNodesMap.entrySet()) {
+            result.putAll(entry.getValue());
+        }
+        return result;
+    }
+
+    @Override
+    public Set<Long> getTimestamps() {
+        return timestampedPathNodesMap.keySet();
+    }
+
+    public void put(Long timestamp, LwM2mPath path, LwM2mNode node) {
```

Comment:
TimestampedLwM2mNodesImpl seems not to be immutable. 
See https://github.com/eclipse/leshan/issues/1192#issuecomment-1058059555


If possible I would prefer to make it immutable.

Reply:
Immutable is TimestampedLwM2mNodes interface. Concrete implementation should not be exposed outside.

Reply:
Not a big deal but I think we don't understand Immutable in a same way. 

I understand that immutable means that an object can not be modified once it is created.
So this is not really applicable to an interface because an interface can not be created.
And in this case, the object behind the interface can be modified  by using cast. 

Anyway, you can let this point aside if you want :-)

Reply:
There are many ways to achieve the goal. I changed the implementation to make TimestampedLwM2mNodes immutable (without interface).